components/core/src/server/oprtenantnode.ts

```diff
@@ -0,0 +1,540 @@
+      nextRunTimestampUTC = result.earliestNextRequestUTC;
+      await this.offerModel.processUpdate(producer.id, result);
+      await this.offerModel.writeOfferProducerMetadata({
+        lastUpdateTimeUTC: now,
+        nextRunTimestampUTC: nextRunTimestampUTC ?? now,
+        organizationUrl: producer.id,
+      });
+      this.logger.info('Ingestion of', producer.id, 'succeeded');
+      this.logger.debug('dbg Lock released on', producer.id);
+    } catch (e) {
+      this.logger.warn('Failed to fetch', producer.id, 'error:', e);
+      nextRunTimestampUTC =
+        now + this.getFailedRetryIntervalMillis(producer.id);
+      await this.offerModel.writeOfferProducerMetadata({
+        lastUpdateTimeUTC: metadata?.lastUpdateTimeUTC,
+        nextRunTimestampUTC: nextRunTimestampUTC ?? now,
+        organizationUrl: producer.id,
+      });
+      this.logger.debug('Lock released on', producer.id);
+    }
+  }
+
+  /**
+   * Returns the minimum number of milliseconds to wait before retrying a
+   * failed request from a producer. This takes the producer id so that
+   * a producer-specific backoff policy can be implemented.
+   */
+  protected getFailedRetryIntervalMillis(producerId: string): number {
+    // TODO: Build a useful, per-producer backoff policy.
+    return 10 * 1000 /* 10 seconds */;
+  }
+
+  private pathToUrl(path?: string): string | undefined {
+    if (path === undefined) {
+      return undefined;
+    }
+    const baseUrl = new URL('./', this.hostOrgUrl);
+    if (path.startsWith('/')) {
+      path = '.' + path;
+    }
+    return new URL(path, baseUrl).toString();
+  }
+
+  private getJwksUrl(): string | undefined {
+    if (!this.jwksURL) {
+      return undefined;
+    }
+    const testUrl = new URL(this.jwksURL, 'http://localhost');
+    if (testUrl.hostname === 'localhost') {
+      return this.pathToUrl('.' + testUrl.pathname);
+    }
+    return this.jwksURL;
+  }
+
+  private getOrgConfig(): OrgConfig {
+    return this.orgConfig;
+  }
+
+  serveOrgFile(): void {
+    const orgFilePath = this.orgFilePath;
+
+    const orgJson = this.getOrgConfig();
+    this.router.get(orgFilePath, async (req, res) => {
+      res.json(orgJson);
+    });
+  }
+
+  maybeServeJwks(): void {
+    if (this.jwksURL) {
+      const testUrl = new URL(this.jwksURL, 'http://localhost');
+      if (testUrl.hostname === 'localhost') {
+        // If the jwksURL is just a pathname, we need to host the
+        // key set.
+        const jwks = this.jwksProvider;
+        if (!jwks) {
+          throw new StatusError(
+            'No jwksProvider specified. A JWKS provider is required if the ' +
+              'frontend server is hosting the JWKS file.',
+            'NO_JWKS_PROVIDER'
+          );
+        }
+        this.router.get(testUrl.pathname, async (req, res) => {
+          res.send(await jwks.getJwks());
+        });
+      }
+    }
+  }
+
+  getRouter(): Router {
+    return this.router;
+  }
+
+  serveApiEndpoints(): void {
```

Comment:
Would we want to allow a server to not host any of the core endpoints? Like could a recipient-only org just provide `/history`, for example?

Reply:
Yes, and that's already allowed. We don't host an endpoint if it doesn't have a path specified in org.json